### PR TITLE
Grading Overview: TS/Tremor -> AG/Blueprint Migration

### LINKED ISSUE #2804: Grading UI/UX: Grading Table is not Responsive
As seen in the image, this is the left most you can go which cuts off the table. On smaller screen sizes, it is even worse.
Screenshot 2024-02-23 at 10 24 00 AM

### LINKED ISSUE #2812: Edge case bugs introduced through transfer of pagination of grading submissions from frontend to backend.
Relevant PRs:
Frontend: https://github.com/source-academy/frontend/pull/2787
Backend: https://github.com/source-academy/backend/pull/1065

The migration of pagination work from the frontend to the backend is a breaking change, and has resulted in the following edge case bugs which are noted here, so as to scope the PRs above. These bugs do not apply for most use cases. They may have arisen due to remaining frontend filtering from tanstack or interactions with the workspace state.

**Page switching with filters on result in frontend filtering taking over**

https://github.com/source-academy/frontend/assets/110712708/da2e6203-581f-4f39-9c21-64b5b412eaaa

Filters, which persist throughout the same session in the workspace, typically fire off backend queries that fill the entire page up. However, if we leave the grading submissions table and revisit it, the frontend filtering takes over just for initialization. Backend-driven behaviour is reinstated after any other action.

Possible ways to fix:
- Under part of a larger pull to remove tanstack/tremor entirely.
- Rewire table initialization logic to account for non-empty filters immediately.
- Feature removal: make filters entirely self-contained and remove it from workspace state

**Mutually exclusive progress column filters and dropdown viewing filters result in frontend filtering taking over**

https://github.com/source-academy/frontend/assets/110712708/ea148c48-c854-4bdb-951e-e1fbb1a4371b

Entries marked as ungraded by our dropdown options must necessarily have progress 'submitted'.  However, filters based on progress can persist in the switch of viewing from 'all' to 'ungraded'. If these filters become mutually exclusive, e.g. viewing ungraded, filter progress 'Attempting', the 'Attempting' filter then performs frontend filtering, while the viewing 'ungraded' option powers the backend querying as expected. This bug does not occur for non-mutually exclusive combinations, like  'Attempted' and 'all' , and 'Submitted' and 'ungraded' (for this case, suspect that frontend filtering still triggers but does not remove any entries within each page)

Possible ways to fix:
- Under part of a larger pull to remove tanstack/tremor entirely.
- Remove filtering by progress if viewing changes from 'All' to 'ungraded'
- Feature removal: make the progress column non filterable.

### PR DESCRIPTION
**Pull request by InfinityTwo**
# Description

This is a major update to the Grading Overview page which migrates from Tanstack Table and Tremor UI to AG Grid and Blueprint JS. It also introduces a few new features that change the user workflow.

Summary
-----------

- **Preferably merged with backend [#1107](https://github.com/source-academy/backend/pull/1107)**
Frontend does not error without it, but sorting columns will not sort anything.

- Part of #2525, #2849 and #2665.

- Resolves #2804 and #2812.

- Adds onto #2856.

> [!NOTE]
> For future Avengers and Developers, we would like you to feedback and think about these aspects for future development
> - XP sorting. Currently sorted by raw XP, but is it preferred over combined XP?
> - Grading/Filter mode. Is it better to default to Grading or Filter mode? Do you have any default preferences? (Devs can consider session storage for this, and column filters)
> - Is it better to sort Type by ordering of the tabs at the top (e.g Missions, Quests, Paths, Contests...) or alphabetically?
> - Is it helpful to add a column for submission date/time?
>
> Please give your feedback in the latest Avenger feedback form/issue tracker or feel free to submit a new issue tracker. Thank you!

# Features/Changes Added

### Feature/Change 1 - Migration of Table
Part of #2525 and #2849. Adds onto #2856.

All components under Grading Overview will have their components migrated from Tanstack Table and Tremor UI to AG Grid and Blueprint JS. This also means some existing features are modified and may not work the same as before. I have tried my best to keep the table looking the same as before, less the changes made as new features. In #2856, Publish and Unpublish grading buttons were added. I've changed it to become icons to better suit the table. The Unpublish icon can be seen in the picture below, with the first row with the grading icon. The publish icon is at the second row of icons, the last icon.

Screenshot_1

### Feature/Change 2 - Grading/Filter Mode

The default workflow is grading mode. Click on the Grading Mode/Filter Mode button beside the search bar.

In Grading mode, the entire row of each submission (except actions) will be clickable, which navigates the user straight to the submission grading page. This was added as the expected workflow for new users (avengers) was to be able to enter the grading of that submission when clicked, instead of filtering it. The Grading action icon will be removed to declutter the table in this mode.

In Filter Mode, it works the same as the previous table. When you click on any cell that allows filtering, it will filter by the value of that cell throughout the table. The Grading action button will be added back here to avoid the friction of going back to Grading Mode.

Besides the button to differentiate Grading/Filter Mode, when you hover over filterable cells, in Grading Mode, individual cells will not be highlighted/underlined when hovered but in Filter Mode, it will do so.

Filter Mode
Screenshot_3

Grading Mode
Screenshot_2

### Feature/Change 3 - Customer Column Headers - Sorting & Hideable Columns
Somewhat part of #2665.

On the top of each column header, there are two buttons on the right. The first button allows you to sort in ascending/descending order. The workflow is don't sort by default, click to sort by ascending, click again to sort by descending, and click again to not sort it.

Note that the sorting is done on the backend with this PR [#1107](https://github.com/source-academy/backend/pull/1107)

The second button allows you to hide the respective columns. This allows you to view the table with less clutter, and it helps with devices that have a smaller screen.

Custom Headers with filtered columns and cell
Screenshot_4

### Other Minor Changes
- (a) Loading Table Overlay

When filtering or searching for new queries, the table will not load in the new data if another query has been made. On the shell, it still queries for the data in the backend and receives every new query like the previous table as this is a purely frontend change.

I have also added a loading animation to show that it's retrieving the requested data. It shows the animation on the first query and hides it when the most recent data from the most recent query has been received.

The table will also become uninteractable (filtering will still be interactable) when querying for the data.

Tl;dr of how it works: when a query is sent, the counter increments by 1. When a query is received, the counter decrements by 1 or stays at 0 if already at 0. If the counter !== 0, then the loading is shown. In certain rare cases, it may not show the loading overlay (still can't figure out why).

a

- (b) Resolves #2804 left cutoff, but the fix is not 100% mobile responsive. It will work on mobile, but the user will need to scroll right.
- (c) Resolves #2812. This introduces a removal of non-"graded" filters when viewing unpublished submissions as that should only show unpublished but graded submissions.
- (d) Added a refresh button at the top of the table to refresh the same query (for convenience in case you want to query an updated database with the same options/filters).

E

# Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
- [ ] Code quality improvements

# How to test

Head over to the Grading page and explore it over there.

# Checklist

- [x] I have migrated the core components & features
- [x] I have cleaned up unrequired files and code
- [x] I have modified mock data
- [x] I have extensively tested this code
- [x] I have successfully merged it with the latest master branch (including P2's isGradingPublished PR) and fixed all conflicts
- [x] Backend Sorting has been completed
- [ ] I have updated the documentation (Not sure if needed)

### CODE REVIEW

File: src/pages/academy/grading/subcomponents/GradingSubmissionsTable.tsx
Comment (RichDom2185) [inline src/pages/academy/grading/subcomponents/GradingSubmissionsTable.tsx:1]: In the resolution of merge conflicts, I went to prioritise your changes over the recently merged PRs, thus the current version does not support team assessments (in fact, some of the other files give a compile error).

Can you help take a look? I've attached the fields that were modified by team assessment PR as a code comment.

The basic idea is that now there will be an XOR constraint (either `studentName` will be populated in case of an individual assessment, or `studentNames` for team assessment – same for `studentUsername` and `studentUsernames`). Thus the table display will fallback to joining the `studentNames`/`studentUsernames` field if the other one is empty.

Refer to #2548 for more details.

Please contact me via Telegram if you require any clarifications.

Thanks!

File: src/commons/grading/GradingFlex.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
+declare const twJustifyContentValues: readonly ["justify-start", "justify-end", "justify-center", "justify-between", "justify-around", "justify-evenly"];
```
Comment (RichDom2185) [inline src/commons/grading/GradingFlex.tsx:1]: Can I check what is this used for?

File: src/commons/grading/GradingFlex.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        : justifyContent === "justify-center"
+        ? "center"
+        : justifyContent === "justify-between"
+        ? "space-between"
+        : justifyContent === "justify-around"
+        ? "space-around"
+        : justifyContent === "justify-evenly"
+        ? "space-evenly"
+        : ""
+      ),
```
Comment (RichDom2185) [inline src/commons/grading/GradingFlex.tsx:34]: Seems like we are mapping Tailwind values to actual styles – what's the need for this? Can we just map input props directly?

File: src/commons/grading/GradingFlex.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  };
+
+  return (
+    <div className={className} style={{...defaultStyle, ...style}}>
+      {children}
+    </div>
+  );
+}
+
+export default GradingFlex;
```
Comment (RichDom2185) [inline src/commons/grading/GradingFlex.tsx:56]: Please add an empty line at the end of this file.

File: src/commons/grading/GradingText.tsx [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    <Text 
+      className={"bp5-ui-text " + className + (secondaryText ? " bp5-text-muted" : "")} 
+      style={{...defaultStyle, ...style}}
+    >
+      {children}
+    </Text>
+  );
+}
+
+export default GradingText;
```
Comment (RichDom2185) [inline src/commons/grading/GradingText.tsx:26]: Ditto on newline at EOF

File: src/commons/sagas/BackendSaga.ts [hunk 5]
```diff
@@ -466,6 +484,7 @@ function* BackendSaga(): SagaIterator {
         gradedFilter,
         pageParams,
         filterParams
+        // sortedBy
```
Comment (RichDom2185) [inline src/commons/sagas/BackendSaga.ts:487]: Is this going to be used/removed?

**Review by RichDom2185 — COMMENTED** [review]
Hi, looking forward for the changes! For now, I have some preliminary comments below:

---

Thanks for working on this!

File: src/commons/sagas/BackendSaga.ts (hunk 5 above)
Comment (InfinityTwo) [reply src/commons/sagas/BackendSaga.ts:487]: It will be used for backend column sorting onced implemented, commented for now in case it causes errors.

File: src/commons/grading/GradingFlex.tsx (hunk 2 above)
Comment (InfinityTwo) [reply src/commons/grading/GradingFlex.tsx:34]: I was just being lazy and trying to avoid touching existing props but I can go through all of it again when I work on it to remove it.

File: src/commons/grading/GradingFlex.tsx (hunk 1 above)
Comment (InfinityTwo) [reply src/commons/grading/GradingFlex.tsx:1]: think I copied it blindly to avoid errors, will look through as well with the other next comment when I get around to looking at it.

File: src/pages/academy/grading/subcomponents/GradingSubmissionsTable.tsx
Comment (InfinityTwo) [reply src/pages/academy/grading/subcomponents/GradingSubmissionsTable.tsx:1]: Thanks for merging up with most of the conflicts, I'll let you know if I need help with this one!

**Comment by josh1248** [comment]
does this change have the edge case bugs in #2812? seems likely that aggrid's way of doing things over tanstack could help to curb the forceful frontend filtering and solve the issue.

**Comment by InfinityTwo** [comment]
> does this change have the edge case bugs in #2812? seems likely that aggrid's way of doing things over tanstack could help to curb the forceful frontend filtering and solve the issue.

It does indeed have that case as the way data is handled is not under aggrid, but I will provide a (rough frontend) fix it in the next commit.

File: src/commons/grading/GradingFlex.tsx (hunk 1 above)
Comment (InfinityTwo) [reply src/commons/grading/GradingFlex.tsx]: This has been fixed in the latest commit, it can be resolved once you are ok with it.

File: src/commons/grading/GradingFlex.tsx (hunk 2 above)
Comment (InfinityTwo) [reply src/commons/grading/GradingFlex.tsx]: This has been fixed in the latest commit, it can be resolved once you are ok with it.

File: src/commons/grading/GradingFlex.tsx (hunk 3 above)
Comment (InfinityTwo) [reply src/commons/grading/GradingFlex.tsx]: This has been fixed in the latest commit, it can be resolved once you are ok with it.

File: src/commons/grading/GradingText.tsx (hunk 4 above)
Comment (InfinityTwo) [reply src/commons/grading/GradingText.tsx]: This has been fixed in the latest commit, it can be resolved once you are ok with it.

File: src/commons/sagas/BackendSaga.ts (hunk 5 above)
Comment (InfinityTwo) [reply src/commons/sagas/BackendSaga.ts]: This has been uncommented in the latest commit and the various issues that comes with it has been fixed.

File: src/pages/academy/grading/subcomponents/GradingSubmissionsTable.tsx
Comment (InfinityTwo) [reply src/pages/academy/grading/subcomponents/GradingSubmissionsTable.tsx:1]: I've merged the changes for this in the latest commit, but I'm unsure if the data shown is correct. I have messaged you on this.

**Comment by josh1248** [comment]
only work needed to resolve the conflict should be at the grading submissions table - there, you would have noticed 1 instead of 2 columns within tanstack. it should hopefully be a straightforward deletion of 1 un-needed column and the renaming of another.

File: src/commons/mocks/BackendMocks.ts [hunk 6]
```diff
@@ -181,9 +181,24 @@ export function* mockBackendSaga(): SagaIterator {
     FETCH_GRADING_OVERVIEWS,
     function* (action: ReturnType<typeof actions.fetchGradingOverviews>): any {
       const accessToken = yield select((state: OverallState) => state.session.accessToken);
-      const { filterToGroup, pageParams, filterParams } = action.payload;
+      const { filterToGroup, pageParams, filterParams, allColsSortStates } = action.payload;
+      const sortedBy = {
+        sortBy: allColsSortStates.sortBy,
+        sortDirection: ''
+      };
+      for (const key in allColsSortStates.currentState) {
```
Comment (RichDom2185) [inline src/commons/mocks/BackendMocks.ts:189]: Is this type safe? How is it different from `Object.keys(allColsSortStates.currentState).forEach(...)`?

File: src/features/grading/GradingTypes.ts [hunk 7]
```diff
@@ -9,6 +11,24 @@ import {
...
+  assessmentName = 'assessmentName',
+  assessmentType = 'assessmentType',
+  studentName = 'studentName',
+  studentUsername = 'studentUsername',
+  groupName = 'groupName',
+  progressStatus = 'progressStatus',
+  gradingStatus = 'gradingStatus',
+  xp = 'xp',
+  actionsIndex = 'actionsIndex'
+}
```
Comment (RichDom2185) [inline src/features/grading/GradingTypes.ts:24]: This might result in coupling; but just to confirm since I'm not too familiar: why must this be a separate definition instead of using TS `keyof`?

File: src/features/grading/GradingTypes.ts [hunk 8]
```diff
@@ -9,6 +11,24 @@ import {
...
+
+export enum SortStates {
+  ASC = 'sort-asc',
+  DESC = 'sort-desc',
+  NONE = 'sort'
```
Comment (RichDom2185) [inline src/features/grading/GradingTypes.ts:29]: If you're using an enum, please use `SortStates.NONE` above (in the backend saga) instead of the raw string value.

File: src/pages/academy/grading/Grading.tsx [hunk 9]
```diff
@@ -48,22 +53,52 @@ const Grading: React.FC = () => {
...
+
   const updateGradingOverviewsCallback = useCallback(
     (page: number, filterParams: Object) => {
+      // Prevents es-lint missing dependency warning
+      if (refreshQueried) {
+        return setRefreshQueried(false);
+      }
+
+      dispatch(setGradingHasLoadedBefore());
+      dispatch(increaseRequestCounter());
```
Comment (RichDom2185) [inline src/pages/academy/grading/Grading.tsx:72]: Can I check what's the request counter for?

**Review by RichDom2185 — COMMENTED** [review]
Hi, thanks for working on this, did a quick run through and some comments/questions below:

File: src/features/grading/GradingTypes.ts (hunk 7 above)
Comment (InfinityTwo) [reply src/features/grading/GradingTypes.ts:24]: I did use the string as values in some of the areas in the table, which ```type``` doesn't allow the use of it (or not that I know of at least)

File: src/pages/academy/grading/Grading.tsx (hunk 9 above)
Comment (InfinityTwo) [reply src/pages/academy/grading/Grading.tsx:72]: Loading Overlay. Also, I've added the relevant changes to the other comments that I didn't comment on in the latest commit 😄

File: src/features/grading/GradingTypes.ts [hunk 10]
```diff
@@ -69,6 +104,69 @@ export type GradingQuery = {
...
+};
+
+export enum ColumnName {
+  assessmentName = 'Name',
+  assessmentType = 'Type',
+  studentName = 'Student(s)',
+  studentUsername = 'Username(s)',
+  groupName = 'Group',
+  progressStatus = 'Progress',
+  xp = 'Raw XP (+Bonus)',
```
Comment (lhw-1) [inline src/features/grading/GradingTypes.ts:122]:
```suggestion
  xp = 'Raw XP (+ Bonus)',
```
Just a very tiny unimportant personal preference, feel free to reject based on whichever looks better in the UI 😅

**Review by lhw-1 — COMMENTED** [review]
First of all, I appreciate the extensive documentation on the PR itself. Great job! 😄 

In terms of the migration & features, I like that the UI was kept consistent despite the migration - thanks for the effort! One additional suggestion here might be to include a tooltip for how Filter Mode vs Grading Mode works, but it's not urgent & can be a separate issue.

I have a few queries / nits on the changes below:

File: src/commons/grading/GradingFlex.tsx [hunk 11]
```diff
@@ -0,0 +1,33 @@
+import { Property } from 'csstype';
+import React from 'react';
+
+const defaultStyles: React.CSSProperties = {
+  display: 'flex'
+};
```
[the remark this answers is not in the document]
Comment (InfinityTwo) [reply src/commons/grading/GradingFlex.tsx:6]: It's a style that will be applied to all GradingFlex/GradingText by default, and separated for clarity too.

File: src/features/grading/GradingTypes.ts [hunk 12]
```diff
@@ -53,6 +74,20 @@ export type GradingOverviewWithNotifications = {
  */
 export type GradingAnswer = GradingQuestion[];
 
+export type AllColsSortStates = {
+  currentState: SortStateProperties;
+  sortBy: ColumnFieldsKeys | '';
```
[the remark this answers is not in the document]
Comment (InfinityTwo) [reply src/features/grading/GradingTypes.ts:79]: The empty string allows us to not sort by any columns, which is the default state from my memory

File: src/features/grading/GradingTypes.ts (hunk 10 above)
Comment (InfinityTwo) [reply src/features/grading/GradingTypes.ts:122]: I think the one without space looks better in the UI after trying it out, having a space feels a little off

File: src/commons/mocks/BackendMocks.ts [hunk 13]
```diff
@@ -166,9 +167,25 @@ export function* mockBackendSaga(): SagaIterator {
...
       const accessToken = yield select((state: OverallState) => state.session.accessToken);
-      const { filterToGroup, pageParams, filterParams } = action.payload;
+      const { filterToGroup, pageParams, filterParams, allColsSortStates } = action.payload;
+      const sortedBy = {
+        sortBy: allColsSortStates.sortBy,
+        sortDirection: ''
+      };
+
+      Object.keys(allColsSortStates.currentState).forEach(key => {
+        if (allColsSortStates.sortBy === key && key != '') {
```
[the remark this answers is not in the document]
Comment (InfinityTwo) [reply src/commons/mocks/BackendMocks.ts]: I've changed it to ```key``` since it works too 😄

**Comment by RichDom2185** [comment]
I'm gonna make some minor improvements, hence marking as draft. Thanks for the work so far!

File: src/pages/academy/grading/subcomponents/GradingBadges.tsx [hunk 14]
```diff
@@ -1,28 +1,102 @@
...
+
+declare const sizeValues: readonly ['xs', 'sm', 'md', 'lg', 'xl'];
+declare type Size = (typeof sizeValues)[number];
+
+type BadgeProps = {
+  text: string;
+  /** First color is bg, second is text. Refer to {typeof AVAILABLE_COLORS} */
+  color?: readonly [string, string];
+  size?: Size;
+  icon?: () => React.ReactNode;
```
Comment (RichDom2185) [inline src/pages/academy/grading/subcomponents/GradingBadges.tsx:18]: Let's just pass the node directly. That way no need for optional chaining as well since `undefined` is a valid React node.

**Review by RichDom2185 — CHANGES_REQUESTED** [review]
Just some minor nits :)

File: src/pages/academy/grading/subcomponents/GradingBadges.tsx (hunk 14 above)
Comment (RichDom2185) [reply src/pages/academy/grading/subcomponents/GradingBadges.tsx]: Resolved in 9bc265690e6cda0a68b470d823b21273a5b90b6d

**Review by RichDom2185 — APPROVED** [review]
LGTM, thanks a lot! Sorry for the delayed review&hellip;
